target/linux/ath79/files/drivers/mtd/nand/raw/rb4xx-nand.c

```diff
@@ -0,0 +1,316 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * NAND driver for the MikroTik RouterBoard 4xx series
+ *
+ * Copyright (C) 2008-2011 Gabor Juhos <juhosg@openwrt.org>
+ * Copyright (C) 2008 Imre Kaloz <kaloz@openwrt.org>
+ * Copyright (C) 2015 Bert Vermeulen <bert@biot.com>
+ * Copyright (C) 2020 Christopher Hill <ch6574@gmail.com>
+ *
+ * This file was based on the driver for Linux 2.6.22 published by
+ * MikroTik for their RouterBoard 4xx series devices.
+ */
+#include <linux/platform_device.h>
+#include <linux/mtd/rawnand.h>
+#include <linux/gpio/consumer.h>
+#include <linux/module.h>
+#include <linux/of_device.h>
+#include <linux/version.h>
+
+#include <mfd/rb4xx-cpld.h>
+
+struct rb4xx_nand {
+	struct rb4xx_cpld *cpld;
+	struct device *dev;
+
+	struct nand_chip chip;
+	struct gpio_desc *ale;
+	struct gpio_desc *cle;
+	struct gpio_desc *nce;
+	struct gpio_desc *rdy;
+};
+
+static int rb4xx_ooblayout_ecc(struct mtd_info *mtd, int section,
+				struct mtd_oob_region *oobregion)
+{
+	switch (section) {
+	case 0:
+		oobregion->offset = 8;
+		oobregion->length = 3;
+		return 0;
+	case 1:
+		oobregion->offset = 13;
+		oobregion->length = 3;
+		return 0;
+	default:
+		return -ERANGE;
+	}
+}
+
+static int rb4xx_ooblayout_free(struct mtd_info *mtd, int section,
+				struct mtd_oob_region *oobregion)
+{
+	switch (section) {
+	case 0:
+		oobregion->offset = 0;
+		oobregion->length = 4;
+		return 0;
+	case 1:
+		oobregion->offset = 4;
+		oobregion->length = 1;
+		return 0;
+	case 2:
+		oobregion->offset = 6;
+		oobregion->length = 2;
+		return 0;
+	case 3:
+		oobregion->offset = 11;
+		oobregion->length = 2;
+		return 0;
+	default:
+		return -ERANGE;
+	}
+}
+
+static const struct mtd_ooblayout_ops rb4xx_nand_ecclayout_ops = {
+	.ecc = rb4xx_ooblayout_ecc,
+	.free = rb4xx_ooblayout_free,
+};
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,20,0)
+static uint8_t rb4xx_nand_read_byte(struct mtd_info *mtd)
+{
+	struct rb4xx_nand *nand = mtd->priv;
+#else
+static u8 rb4xx_nand_read_byte(struct nand_chip *chip)
+{
+	struct rb4xx_nand *nand = chip->priv;
+#endif
+
+	struct rb4xx_cpld *cpld = nand->cpld;
+	u8 data;
+	int ret;
+
+	ret = cpld->rb4xx_cpld_read_nand(cpld, &data, 1);
+	if (unlikely(ret)) {
+		dev_err(nand->dev, "cpld read nand failed: %d\n", ret);
```

Comment:
I'll say this only here but I think you're adding printks where they don't belong. These read/write functions should be slim and (very) fast. It's not their job to report errors in human-readable way. Plus any problem here will flood dmesg with the chance of obliterating helpful debug info. I'd remove these and likewise in gpio. Printing out error is the job of higher layers.

Reply:
I've removed them - they were originally added when I was testing things out, so no longer needed and I agree on flooding the console.